File: algorithms/11_golang/data_structures/monotonic_stack.go
package main

import "fmt"

func nextGreater(nums []int) []int {
	n := len(nums)
	res := make([]int, n)
	for i := range res {
		res[i] = -1
	}
	stack := make([]int, 0)
	for i := 0; i < n; i++ {
		for len(stack) > 0 && nums[stack[len(stack)-1]] < nums[i] {
			top := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			res[top] = nums[i]
		}
		stack = append(stack, i)
	}
	return res
}

func main() {
	nums := []int{2, 1, 2, 4, 3}
	got := nextGreater(nums)
	want := []int{4, 2, 4, -1, -1}
	for i := range want {
		if got[i] != want[i] {
			panic("monotonic stack: next greater mismatch")
		}
	}
	fmt.Println("[Go MonotonicStack] Next greater element verified:", got)
}

File: algorithms/03_cpp/data_structures/monotonic_stack.cpp
/**
 * Monotonic stack: next greater element for every position.
 */

#include <vector>
#include <stack>
#include <cassert>
#include <iostream>

std::vector<int> nextGreater(const std::vector<int>& a) {
    int n = static_cast<int>(a.size());
    std::vector<int> res(n, -1);
    std::stack<int> st;
    for (int i = n - 1; i >= 0; --i) {
        while (!st.empty() && st.top() <= a[i]) st.pop();
        res[i] = st.empty() ? -1 : st.top();
        st.push(a[i]);
    }
    return res;
}

std::vector<int> nextGreaterIndex(const std::vector<int>& a) {
    int n = static_cast<int>(a.size());
    std::vector<int> res(n, -1);
    std::stack<int> st;
    for (int i = n - 1; i >= 0; --i) {
        while (!st.empty() && a[st.top()] <= a[i]) st.pop();
        res[i] = st.empty() ? -1 : st.top();
        st.push(i);
    }
    return res;
}

int main() {
    std::vector<int> a = {2, 1, 5, 6, 2, 3};
    auto ng = nextGreater(a);
    assert((ng == std::vector<int>{5, 5, 6, -1, 3, -1}));
    auto ngi = nextGreaterIndex(a);
    assert((ngi == std::vector<int>{2, 2, 3, -1, 5, -1}));
    std::cout << "[C++ MonotonicStack] Next greater element verified." << std::endl;
    return 0;
}
